### PR TITLE
PLUGIN-461: fix issue with backward incompatibility

### PR DESCRIPTION
**Pull request by ardian4**
1. if property ``serviceAccountType`` is null and is not macro fallback to ``filePath``
2. Handle macro for serviceAccountType

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java [hunk 1]
```diff
@@ -90,10 +90,10 @@ public String getServiceAccountJson() {
   }
 
   public String getServiceAccountType() {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java:96]: If this returns null the method should be marked as nullable. Also make sure all the callers handle null values

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java [hunk 2]
```diff
@@ -107,8 +107,10 @@ public static String extractPrivateKey(ServiceAccountCredentials credentials) {
     boolean isServiceAccountFilePath = SERVICE_ACCOUNT_TYPE_FILE_PATH.equals(serviceAccountType);
 
     for (String prefix : keyPrefix) {
-      if (isServiceAccountFilePath && serviceAccount != null) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:110]: can it still be null if its a macro?

**Review by CuriousVini — COMMENTED** [review]
- Can we add a test for this case?
- We observed similar behavior in GCS Source. Has it been fixed?
https://issues.cask.co/browse/PLUGIN-461?focusedCommentId=39830&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-39830

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java (hunk 2 above)
Comment (ardian4) [reply src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:110]: `isServiceAccountFilePath` will be false when `serviceAccountType` is macro, but code is updated so when `serviceAccountType` is macro, it will not add any auth properties.

**Comment by ardian4** [comment]
> * Can we add a test for this case?
> * We observed similar behavior in GCS Source. Has it been fixed?
>   https://issues.cask.co/browse/PLUGIN-461?focusedCommentId=39830&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-39830

- Tests added 
- Also issue with GCSSource was already fixed in previous commit 

https://github.com/data-integrations/google-cloud/blob/96509a39ecec83cea8e581d1041e85d3683acc0e/src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java#L109-L115

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java (hunk 2 above)
Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:110]: So basically below check is exactly same as this one. Then why do we need that change?

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java [hunk 3]
```diff
@@ -101,14 +101,19 @@ public static String extractPrivateKey(ServiceAccountCredentials credentials) {
                                                            String serviceAccountType,
                                                            String... keyPrefix) throws IOException {
     Map<String, String> properties = new HashMap<>();
+    if (serviceAccountType == null) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:104]: Why is this change added? why it was not needed before?

File: src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java [hunk 4]
```diff
@@ -89,18 +89,21 @@ public String getServiceAccountJson() {
...
   }
 
+  @Nullable
   public String getServiceAccountType() {
-    if (containsMacro(NAME_SERVICE_ACCOUNT_TYPE) || Strings.isNullOrEmpty(serviceAccountType)) {
+    if (containsMacro(NAME_SERVICE_ACCOUNT_TYPE)) {
       return null;
     }
-    return serviceAccountType;
+    return Strings.isNullOrEmpty(serviceAccountType) ? SERVICE_ACCOUNT_FILE_PATH : serviceAccountType;
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/common/GCPConfig.java:100]: Please add a comment in the javadoc explaining what is the default value that is interpreted

**Review by CuriousVini — COMMENTED** [review]
Are these changes applicable to any of the other GCP plugins such as pubsub, speechtransform?

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java (hunk 2 above)
Comment (ardian4) [reply src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:110]: This change is required for the case when serviceAccountType is filePath and serviceAccount is null - for autodetect scenario. If we leave as it was it would fallback back to serviceAccountType as it was a JSON key.

File: src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java (hunk 3 above)
Comment (ardian4) [reply src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java:104]: It was added to handle when serviceAccountType is macro, and this method is called in configuration stage.

**Comment by ardian4** [comment]
> Are these changes applicable to any of the other GCP plugins such as pubsub, speechtransform?

Not all plugins use this method. pubsub and speechtransform are not using this method.

**Review by CuriousVini — APPROVED** [review]
lgtm. Please squash the commits

**Comment by ardian4** [comment]
> lgtm. Please squash the commits

Squashed

**Review by CuriousVini — APPROVED** [review]
